src/Type.hs

```diff
@@ -221,69 +257,84 @@ checkAssign _ _ = error "todo: support struct assignment in constructors"
 
 -- ensures key and value types match when assigning a defn to a mapping
 -- TODO: handle nested mappings
-checkDefn :: Env -> AbiType -> AbiType -> Id -> U.Defn -> Err StorageUpdate
-checkDefn env@Env{contract} keyType (FromAbi valType) name (U.Defn k val) =
+checkDefn :: Pn -> Env -> ValueType -> ValueType -> Id -> U.Defn -> Err StorageUpdate
+checkDefn pn env@Env{contract} keyType vt@(FromVType valType) name (U.Defn k val) =
   _Update
-  <$> (_Item contract name <$> checkIxs env (getPosn k) [k] [keyType])
+  <$> (_Item vt . SMapping nowhere (SVar pn contract name) <$> checkIxs env (getPosn k) [k] [keyType])
   <*> checkExpr env valType val
 
 -- | Typechecks a postcondition, returning typed versions of its storage updates and return expression.
 checkPost :: Env -> U.Post -> Err ([Rewrite], Maybe (TypedExp Timed))
-checkPost env@Env{contract,theirs} (U.Post storage extStorage maybeReturn) = do
+checkPost env@Env{contract,theirs} (U.Post storage maybeReturn) = do
   returnexp <- traverse (typedExp $ focus contract) maybeReturn
-  ourStorage <- checkEntries contract storage
-  otherStorage <- checkStorages extStorage
-  pure (ourStorage <> otherStorage, returnexp)
+  storage' <- checkEntries contract storage
+  pure (storage', returnexp)
   where
     checkEntries :: Id -> [U.Storage] -> Err [Rewrite]
     checkEntries name entries = for entries $ \case
       U.Constant loc     -> Constant <$> checkPattern     (focus name) loc
       U.Rewrite  loc val -> Rewrite  <$> checkStorageExpr (focus name) loc val
 
-    checkStorages :: [U.ExtStorage] -> Err [Rewrite]
-    checkStorages [] = pure []
-    checkStorages (U.ExtStorage name entries:xs) = mappend <$> checkEntries name entries <*> checkStorages xs
-    checkStorages _ = error "TODO: check other storages"
-
     focus :: Id -> Env
     focus name = env
       { contract = name
       , store    = Map.findWithDefault mempty name theirs
       }
 
+checkEntry :: forall t. Typeable t => Env -> U.Entry -> Err (SlotType, StorageRef t)
+checkEntry Env{contract,store,calldata} (U.EVar p name) = case (Map.lookup name store, Map.lookup name calldata) of
+  (Just _, Just _) -> throw (p, "Ambiguous variable " <> name)
+  (Just typ, Nothing) -> pure (typ, SVar p contract name)
+  (Nothing, Just _) -> throw (p, "Variable " <> name <> " is not a storage variable")
+  (Nothing, Nothing) -> throw (p, "Unknown storage variable " <> show name)
+  -- TODO more consitent check of name overlap between calldata and storage
```

Comment:
wdym here?

Reply:
It refers to the line above: 
https://github.com/ethereum/act/blob/26525a8b8c11293b34dc7a1e05b2e46ca3361658/src/Type.hs#L286  

I think it would be better to check whether there are argument names that overlap with state variables in a different pass, rather than having this check whenever we lookup a name since there isn't a way to disambiguate the reference. wdyt?